src/flint/fmpz.jl

```diff
@@ -1139,19 +1139,25 @@ end
 issquare(x::fmpz) = Bool(ccall((:fmpz_is_square, :libflint), Cint,
                                (Ref{fmpz},), x))
 
+isprime(x::UInt) = Bool(ccall((:n_is_prime, :libflint), Cint, (UInt,), x))
```

Comment:
Maybe `Culong` instead of `UInt`? (or is it always the same type?)

Reply:
n_is_prime takes a UInt, not a Culong. Feel free to use GMP.limb or whatever Keno suggested.

Reply:
Won't start changing this here.

Reply:
> n_is_prime takes a UInt, not a Culong

The signature of this function, as found in the docs and in the C file, is `int n_is_prime(mp_limb_t n)`, and `mp_limb_t` is an alias for `ulong`. In the header file one can find `FLINT_DLL int n_is_prime(ulong n);`. So yes `n_is_prime` takes `Culong`, which is Julia's name for C's `ulong`.

My question was due to my ignorance of whether `Culong` can differ from `UInt`, mainly on windows, as I have been bitten by this kind of assumption.

Reply:
ulong is not a C type. Culong is C's unsigned long [1], which is NOT what UInt or mp_limb_t are. So we certainly should not be changing UInt's to Culong, as that would be incorrect. Flint expects mp_limb_t, not unsigned long.

[1] https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/

Reply:
I believed indeed that `ulong` is not an official C type, but it's very frequently aliased to `unsigned long`, actually `ulong` is provided as an alias to `unsigned long` in the `<sys/types.h>` by the gcc and clang compilers. I really didn't expect that flint would redefine its own `ulong`. Sorry for the noise.

Reply:
> Feel free to use GMP.limb 

On the other hand, `GMP.Limb` is usually a typedef for `unsigned long`, so this can be 32 bits on Windows 64, and hence `GMP.Limb != UInt`.

Reply:
That doesn't make much sense. GMP.Limb should be the type for a GMP limb, which should be an mp_limb_t. If that's not the case, then I would suggest there is a fairly significant bug.

Reply:
> GMP.Limb should be the type for a GMP limb, which should be an mp_limb_t

It's exactly what it is. Making sure that there is no bug there was one of my first contribution to Julia in 2014. Before, the `GMP` module was using `Culong` (which is definitely not `UInt`) for `mp_limb_t`, so I introduced the `GMP.Limb` type which checks at build time the number of bits in a limb, by calling out to "libgmp". Then at `__init__` time, we check again that there is no mismatch. There was at least one case (which was the reason for the introduction of `GMP.Limb`, on Windows) where `UInt == Limb != Culong`. 

That said, I believe that `GMP.Limb` is usually equal to `Culong`, unless `_LONG_LONG_LIMB` is defined, but it's hard (for me) to say when this macro is defined. I don't have a windows 64 machine, but maybe someone can try `Base.GMP.Limb == UInt` on this platform. 

Now, whether Flint's `mp_limb_t` matches GMP's `mp_limb_t`, I can't tell. But `GMP.Limb` is always correct, with respect to "libgmp", AFAICT.

Reply:
Flint doesn't have an mp_limb_t. That is GMP's mp_limb_t, which is why we should be using either UInt or GMP.limb and not Culong.

Reply:
I got it for `Culong`, but I don't know the details of how `Nemo` is built, so was not sure whether the "gmp.h" file included by flint is the same (as in, the macros determining `mp_limb_t` are the same) as the one used by Julia. In any case let's hope that it's the case and that `UInt == GMP.Limb`.

Reply:
It is the same, as Flint actually works with GMP itself. It wraps GMP with a compatibility layer so it works on Windows 64, but it is still the ordinary GMP.